patches/server/0739-Add-PlayerCompostItemEvent.patch

```diff
@@ -0,0 +1,119 @@
++    }
++
++    static BlockState addItem(BlockState iblockdata, LevelAccessor generatoraccess, BlockPos blockposition, ItemStack itemstack, double rand, @Nullable Entity entity) {
++        // Paper end
+         // CraftBukkit end
+         int i = (Integer) iblockdata.getValue(ComposterBlock.LEVEL);
+         float f = ComposterBlock.COMPOSTABLES.getFloat(itemstack.getItem());
+ 
+-        if ((i != 0 || f <= 0.0F) && rand >= (double) f) {
++        // Paper start
++        com.destroystokyo.paper.event.entity.CompostItemEvent event = null;
++        if (generatoraccess != org.bukkit.craftbukkit.util.DummyGeneratorAccess.INSTANCE) {
++            boolean willRaiseLevel = !((i != 0 || f <= 0.0F) && rand >= (double) f); // taken from if conditional below
++            event = new com.destroystokyo.paper.event.entity.CompostItemEvent(entity == null ? null : entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(generatoraccess, blockposition), itemstack.getBukkitStack(), willRaiseLevel);
++        }
++
++        if (event != null && !event.callEvent()) {
++            return null;
++        }
++
++        if (event != null && !event.willRaiseLevel()) {
++            // Paper end
+             return iblockdata;
+         } else {
+             int j = i + 1;
+@@ -407,6 +439,7 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
+         private final LevelAccessor level;
+         private final BlockPos pos;
+         private boolean changed;
++        private BlockState composterState; // Paper
+ 
+         public InputContainer(BlockState state, LevelAccessor world, BlockPos pos) {
+             super(1);
+@@ -428,7 +461,8 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
+ 
+         @Override
+         public boolean canPlaceItemThroughFace(int slot, ItemStack stack, @Nullable Direction dir) {
+-            return !this.changed && dir == Direction.UP && ComposterBlock.COMPOSTABLES.containsKey(stack.getItem());
++            this.composterState = ComposterBlock.addItem(this.state, this.level, this.pos, stack); // Paper
++            return !this.changed && dir == Direction.UP && ComposterBlock.COMPOSTABLES.containsKey(stack.getItem()) && this.composterState != null; // Paper
+         }
+ 
+         @Override
+@@ -442,9 +476,13 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
+ 
+             if (!itemstack.isEmpty()) {
+                 this.changed = true;
+-                BlockState iblockdata = ComposterBlock.addItem(this.state, this.level, this.pos, itemstack);
++                // Paper start
++                if (this.composterState == null) {
++                    return;
++                }
+ 
+-                this.level.levelEvent(1500, this.pos, iblockdata != this.state ? 1 : 0);
++                this.level.levelEvent(1500, this.pos, this.composterState != this.state ? 1 : 0);
++                // Paper end
```

Comment:
Hmm yeah I forgot about this situation when I was making my example patch...

I don't really like this whole moving the add item to the canPlace... method and would like to leave hopper behavior alone. Taking this into account, having two separate events, EntityCompostItemEvent (is cancellable) and CompostItemEvent (is not cancellable), might actually be better. Can have a note that if they want the hopper input stopped altogether, to use one of the hopper events, not this.

Some more input from others would be good here too.